### PR TITLE
v2: Add ability to load without model configuration files and add more examples

### LINKED ISSUE #4617: V2 Triton Server: Deployment of a model with custom execution environment via PVC
## Describe the bug

Can't copy a custom execution environment file to the Triton server when trying to deploy a model via a PVC.
How to set the EXECUTION_ENV_PATH parameter in the config.pbtxt?

## To reproduce

Hey, 
I'm trying to deploy a python model to a Triton servery via a PVC
1. For this im following [this tutorial](https://docs.seldon.io/projects/seldon-core/en/v2/contents/examples/k8s-pvc.html), with the adaptation of using deploying a Triton instead of a ML-server.
2. I can successfully deploy and use one of the Triton example models, to be specific add20
3. For the actual model i want to deploy i need some additional packages, so i followed the [Triton tutorial on how to add a  custom execution environment](https://github.com/triton-inference-server/python_backend/blob/main/README.md#creating-custom-execution-environments)

Now I can't figure out how to set the EXECUTION_ENV_PATH parameter in the config.pbtxt such that the custom execution environment file (e.g. conda-pack.tar.gz) can be copied by rclone and is also detected by the triton server?

As always, thank you very much for your help!

## Expected behaviour

The custom execution environment file should be copied to the server and the model should be deployed successfully. 

## Environment

Proivder: Rancher Desktop
Architekture: ARM64 / M1 Mac
Kubernetes Cluster Version:
Client Version: v1.25.2
Kustomize Version: v4.5.7
Server Version: v1.25.4+k3s1
Deployed Seldon System Images: image: docker.io/seldonio/seldonv2-controller:latest

## Model Details 

``` 
name: "add20"
backend: "python"

input [
  {
    name: "INPUT"
    data_type: TYPE_FP32
    dims: [ 4 ]
  }
]
output [
  {
    name: "OUTPUT"
    data_type: TYPE_FP32
    dims: [ 4 ]
  }
]

instance_group [{ kind: KIND_CPU }]

parameters: {
  key: "EXECUTION_ENV_PATH",
  value: {string_value: "???/conda-pack.tar.gz"}
}
```

model.yaml
``` yaml
apiVersion: mlops.seldon.io/v1alpha1
kind: Model
metadata:
  name: add20
  namespace: seldon-mesh
spec:
  storageUri: "/var/models/add20"
  requirements:
  - python
  - pvc
```

server.yaml
```` yaml
apiVersion: mlops.seldon.io/v1alpha1
kind: Server
metadata:
  name: triton-pvc
  namespace: seldon-mesh
spec:
  serverConfig: triton
  extraCapabilities:
  - "pvc"
  podSpec:
    volumes:
    - name: models-pvc
      persistentVolumeClaim:
        claimName: ml-models-pvc
    containers:
    - name: rclone
      volumeMounts:
      - name: models-pvc
        mountPath: /var/models

```

### PR DESCRIPTION
**Pull request by cliveseldon**
* Adds ability to load some MLServer models without `model-settings.json` file
 * Add ability to load some Triton models without `config.pbtxt` file
 * Add auto-generated `model-settings.json` and `config.pbtxt` files based on Model resource requirements
 * Add a model-zoo notebook with many examples
 * Update inference artifact docs to help users create the appropriate resources for their model type
 * Copies all files from chosen triton folder to model folder to fix #4617 

FIxes: #4617

### CODE REVIEW

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+func createSettingsFile(path string, modelSpec *scheduler.ModelSpec) error {
+	modelSettings, err := getModelSettings(path, modelSpec)
+	if err != nil {
+		return err
+	}
+	return saveModelSettings(path, modelSettings)
+}
+
+func getModelSettings(path string, modelSpec *scheduler.ModelSpec) (*ModelSettings, error) {
+	for _, requirement := range modelSpec.Requirements {
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:31]: `Requirements` is a free-form text for labels. I am not sure if we should rely on them exactly. Perhaps consider having a type for the model? maybe the `cli` can help populating the type?

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/seldonio/seldon-core/apis/go/v2/mlops/scheduler"
+)
+
+var (
+	sklearnExtensions  = []string{"joblib", "pkl", "pickle"}
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:16]: given that mlserver is [very strict](https://github.com/SeldonIO/MLServer/blob/3a81135aa1da96cc0b0c5748ce2ad7ad540c0d3b/runtimes/sklearn/mlserver_sklearn/sklearn.py#L23) about the full name of the model file, should we also try to match here with `model.joblib` etc?

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func createAlibiExplainModelSettings(path string) (*ModelSettings, error) {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			parallelWorkers := 0
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:142]: perhaps add a note in code on why parallel workers is always set to 0 for explainers?

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+func createLightGBMModelSettings(path string) (*ModelSettings, error) {
+	modelUri, err := findModelUri(path, lightgbmExtensions)
+	if err != nil {
+		return nil, err
+	}
+	return createModelSettingsFromUri(modelUri, "mlserver_lightgbm.LightGBMModel"), nil
+}
+
+func createAlibiDetectModelSettings() (*ModelSettings, error) {
+	return &ModelSettings{
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:118]: could we use `createModelSettingsFromUri` helper?

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		Implementation: "mlserver_alibi_detect.AlibiDetectRuntime",
+		Parameters: &ModelParameters{
+			Uri: "./",
+		},
+	}, nil
+}
+
+func createMLFlowModelSettings() (*ModelSettings, error) {
+	return &ModelSettings{
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:127]: ditto

File: scheduler/pkg/agent/repository/mlserver/settings.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+func createAlibiExplainModelSettings(path string) (*ModelSettings, error) {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			parallelWorkers := 0
+			return &ModelSettings{
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings.go:143]: ditto

File: scheduler/pkg/agent/repository/mlserver/settings_test.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+				mlserverHandler := NewMLServerRepositoryHandler(logrus.New())
+				ms, err := mlserverHandler.loadModelSettingsFromFile(filepath.Join(path, mlserverConfigFilename))
+				g.Expect(err).To(BeNil())
+				g.Expect(ms).To(Equal(test.expected))
+			}
+		})
+	}
+}
+
+func TestCreateSKLearnModelSettings(t *testing.T) {
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings_test.go:152]: we are also testing sklearn models above, why this test is different? should we have all sklearn test cases here?

File: scheduler/pkg/agent/repository/mlserver/settings_test.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	"path/filepath"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"github.com/otiai10/copy"
+	"github.com/seldonio/seldon-core/apis/go/v2/mlops/scheduler"
+	"github.com/sirupsen/logrus"
+)
+
+func TestCreateSettingsFile(t *testing.T) {
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/mlserver/settings_test.go:16]: add a test for mlflow, alibi-detect and alibi-explain ?

File: scheduler/pkg/agent/repository/triton/triton.go [hunk 9]
```diff
@@ -42,35 +42,91 @@ func NewTritonRepositoryHandler(logger log.FieldLogger) *TritonRepositoryHandler
...
-	s.Name = modelName
-	data, err := prototext.Marshal(s)
+	return os.WriteFile(path, data, fs.ModePerm)
+}
+
+func (t *TritonRepositoryHandler) createConfigFileWithName(modelName string, path string) error {
+	config := pb.ModelConfig{}
+	config.Name = modelName
+	return saveConfigFile(path, &config)
+}
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/triton/triton.go:118]: question: are we relying on triton detecting the correct backend?

File: scheduler/pkg/agent/repository/triton/triton_test.go [hunk 10]
```diff
@@ -50,51 +51,64 @@ func TestFindModelVersionFolder(t *testing.T) {
...
+			error:          false,
 		},
 		{
-			name:    "NoVersionFolders",
-			folders: []string{"x"},
-			found:   false,
+			name:           "NoVersionFolders",
+			folders:        []string{},
+			found:          false,
+			expectedFolder: "hash",
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/triton/triton_test.go:75]: I dont understand this test case? is it when the user is not specifying the version and not in the correct triton folder structure? why `hash` expected folder?

File: scheduler/pkg/agent/repository/triton/triton_test.go [hunk 11]
```diff
@@ -257,3 +272,73 @@ output [
...
+		{
+			name:                 "files and folders",
+			rcloneFiles:          []string{"t"},
+			rcloneFolders:        []string{"folder"},
+			expectedRepoPaths:    []string{"t"},
+			notExpectedRepoPaths: []string{"folder"},
+		},
+		{
+			name:                 "files and folders with config.pbtxt and existing folders in repo",
+			rcloneFiles:          []string{"t", "config.pbtxt"},
```
Comment (sakoush) [inline scheduler/pkg/agent/repository/triton/triton_test.go:303]: why there is a test case with `config.pbtxt` in  `TestCopyNonConfigFilesToModelRepo`?

**Review by sakoush — COMMENTED** [review]
Lovely PR! 

One main comment is whether using the free form requirements/labels for generating `model-settings.json` can be made better, i.e. detecting file types instead?

File: docs/source/contents/getting-started/index.md [hunk 12]
```diff
@@ -4,6 +4,11 @@
  * [Install locally with Docker Compose](./docker-installation/index.md)
  * [Install onto a Kubernetes cluster](./kubernetes-installation/index.md)
 
+Once installed:
+
+  * Try the existing [examples](../examples/index.md).
+  * Save your own [model artifact](../models/inference-artifacts/index.html#saving-model-artifacts) to test.
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply docs/source/contents/getting-started/index.md:10]: will reword to 
```
Train and deploy your own [model artifact](../models/inference-artifacts/index.html#saving-model-artifacts).
```

File: samples/scripts/models/Makefile [hunk 13]
```diff
@@ -1,23 +1,126 @@
...
+.PHONY: download-mnist-onnx
+download-mnist-onnx:
+	wget https://github.com/onnx/models/raw/main/vision/classification/mnist/model/mnist-12.onnx -O mnist-onnx/model.onnx
+
+.PHONY: upload-mnist-onnx
+upload-mnist-onnx:
+	gsutil cp mnist-onnx/model.onnx gs://seldon-models/scv2/samples/${MLSERVER_FOLDER}/mnist-onnx/model.onnx
+
+#
+# CIFAR10 Tensorflow
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply samples/scripts/models/Makefile:94]: I don't think hdf5 is directly supported by Triton

File: samples/scripts/models/Makefile [hunk 14]
```diff
@@ -1,23 +1,126 @@
...
+.PHONY: mnist-pytorch
+mnist-pytorch: env
+	wget https://raw.githubusercontent.com/pytorch/examples/main/mnist/main.py -O mnist-pytorch/main.py
+	sed -i  's/torch.save(model.state_dict(), "mnist_cnn.pt")/model_scripted = torch.jit.script(model);model_scripted.save("model.pt")/g' mnist-pytorch/main.py
+	cd mnist-pytorch && ../.env/bin/python3 main.py --save-model --epochs 1
+
+.PHONY: upload-mnist-pytorch
+upload-mnist-pytorch:
+	gsutil cp mnist-pytorch/model.pt gs://seldon-models/scv2/samples/${TRITON_FOLDER}/mnist-pytorch/model.pt
+	gsutil cp mnist-pytorch/config.pbtxt gs://seldon-models/scv2/samples/${TRITON_FOLDER}/mnist-pytorch/config.pbtxt
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply samples/scripts/models/Makefile:126]: no

File: scheduler/pkg/agent/repository/mlserver/mlserver.go [hunk 15]
```diff
@@ -89,50 +90,63 @@ type ModelParameters struct {
...
 		} else {
 			// Find highest numbered version folder
 			mvp, err = m.findHighestVersionInPath(path)
 			if err != nil {
-				return "", err
+				return "", false, err
 			}
 		}
 	}
+	// Return top level folder if we can't find a model-settings.json and are not looking for a specific version
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/mlserver/mlserver.go:137]: You can specify it in the Model resource

File: scheduler/pkg/agent/repository/mlserver/settings.go (hunk 1 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/mlserver/settings.go:31]: This is a bigger discussion as all scheduling is based on matching whats specified in the requirements in the model with the capabilities of the server when started and defined in the ServerConfig

File: scheduler/pkg/agent/repository/mlserver/settings.go (hunk 2 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/mlserver/settings.go:16]: I think this only applies if the URI provided in the model-settings.json is a folder. In this case as we are working backwards and creating the model-settings.json we can be more lenient. see [here](https://github.com/SeldonIO/MLServer/blob/3a81135aa1da96cc0b0c5748ce2ad7ad540c0d3b/mlserver/utils.py#L30).

File: scheduler/pkg/agent/repository/mlserver/settings_test.go (hunk 7 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/mlserver/settings_test.go:152]: I think instead will add more tests for each type of artifact

File: scheduler/pkg/agent/repository/triton/triton.go (hunk 9 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/triton/triton.go:118]: yes for types Triton can do this

File: scheduler/pkg/agent/repository/triton/triton_test.go (hunk 10 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/triton/triton_test.go:75]: we allow a user to just put their files in the top level folder. Which is not valid for Triton which always assumes versioned folders. The hash is just the pretend hash that rclone would create. Will add a comment.

File: scheduler/pkg/agent/repository/triton/triton_test.go (hunk 11 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/triton/triton_test.go:303]: Want to check its not copied

File: scheduler/pkg/agent/repository/mlserver/settings.go (hunk 3 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/repository/mlserver/settings.go:133]: ok will do

File: samples/scripts/models/Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (sakoush) [reply samples/scripts/models/Makefile:94]: Perhaps we need to mention it in the docs somewhere that for tensorflow models it has to be in `SavedModel` format if not already there?

File: samples/scripts/models/Makefile (hunk 14 above)
[the remark this answers is not in the document]
Comment (sakoush) [reply samples/scripts/models/Makefile:126]: could you please add a link to the issue/thread discussing why it is not possible?

File: scheduler/pkg/agent/repository/mlserver/settings.go (hunk 1 above)
Comment (sakoush) [reply scheduler/pkg/agent/repository/mlserver/settings.go:31]: I agree but the servers / models can have any set of labels for this. So not necessarily that these tags remain the same in the future and for customer. 

Yeah we could argue that it is out of scope of this PR

**Review by sakoush — APPROVED** [review]
LGTM

File: samples/scripts/models/Makefile (hunk 14 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply samples/scripts/models/Makefile:126]: https://github.com/pytorch/pytorch/issues/38273

[the remark this answers is not in the document]
Comment (cliveseldon) [reply samples/scripts/models/Makefile:126]: will add to Makefile

File: samples/scripts/models/Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (cliveseldon) [reply samples/scripts/models/Makefile:94]: will add
